arts.php

```diff
@@ -150,7 +150,12 @@
         // Prevent the script from timing out
         set_time_limit(0);
 
-        $image = Art::get_from_source($_SESSION['form']['images'][$image_id], 'album');
+        $show_song_art = AmpConfig::get('show_song_art') ?: false;
+        if ($show_song_art) {
```

Comment:
Uhhhhhhhh that's... uhhhhhh

This basically means:
If there's ANYTHING else than `0; "0"; ""; null` in the variable, we get `$type = 'song';`.
Even if we would have `"false"` in it, it would be true.

please take a look at this: https://www.php.net/manual/en/types.comparisons.php

We need a stricter comparison here. :)

Same is further down. I guess your work is not done yet^^

Reply:
Oha! I didn't know that, sorry. Of course this is a very bad style. How can I change the code properly? `if ($show_song_art == "true") {` should work then, right? But then it is a string comparison. Is it possible somehow to convert `$show_song_art` into a correct boolean value?

By the way, I orientated myself in the existing code. It seems that there are a lot of more code snippets with this mistake. As an example have a look at `arts.class.php`, line no. 375: 
`if (AmpConfig::get('write_id3_art')) {`
`           if ($this->type == 'album') {`

Is there anything different to my change that I am missing? Or does this snippet simply have the same problem and should be reviewed within the redevelopment process?

Reply:
> Oha! I didn't know that, sorry. Of course this is a very bad style. How can I change the code properly? if ($show_song_art == "true") { should work then, right? But then it is a string comparison. Is it possible somehow to convert $show_song_art into a correct boolean value?

We don't have to convert something. We can simply use a strict comparison like: `$show_song_art === "true"`

This way we don't have to struggle with how variables cast to the different types as PHP compares the types 'string' and the content.

All this casting stuff is also extremely confusing for me. That's why I always avoid it, if possible. If I see it right, you only get strings back from `AmpConfig::get` so a strict string comparison is the way to go. This way we son't have any problems with miscasts and so on.

You also could simply test it by setting the option to something else than a boolean-string-thingy in the config file. Or try what happens if you set the option to false / "false" ; true / "true" ; 0 / "0" ; 1 /  "1" or bogus.

> By the way, I orientated myself in the existing code. It seems that there are a lot of more code snippets with this mistake. As an example have a look at arts.class.php, line no. 375:
if (AmpConfig::get('write_id3_art')) {
if ($this->type == 'album') {

> Is there anything different to my change that I am missing? Or does this snippet simply have the same problem and should be reviewed within the redevelopment process?

Yepp, seems to have the same problem. But I'm not completely sure about that, as I didn't test it.

Also, as I'm not an expert, everything that I said can be total BS :D

But I think I got it right^^

Reply:
Sorry to bother you again, but I'm afraid you seem to be wrong here. I was not able to get it running with the changes we discussed. So I tested around a bit. In my opinion the `AmpConfig::get` in fact does a convertion into a boolean value. 

I tested the following expressions:

1. Having configured the setting to "true"
- `$show_song_art` === "true"` --> result FALSE
- `$show_song_art` == "true"` --> result FALSE
- `$show_song_art` === 'true'` --> result FALSE
- `$show_song_art` == 'true'` --> result FALSE
- `$show_song_art` == true` --> result TRUE

2. Having configured the setting to "false"
- `$show_song_art` === "true"` --> result FALSE
- `$show_song_art` == "true"` --> result FALSE
- `$show_song_art` === 'true'` --> result FALSE
- `$show_song_art` == 'true'` --> result FALSE
- `$show_song_art` == true` --> result FALSE

3. Having configured the setting to any rubbish
- `$show_song_art` === "true"` --> result FALSE
- `$show_song_art` == "true"` --> result FALSE
- `$show_song_art` === 'true'` --> result FALSE
- `$show_song_art` == 'true'` --> result FALSE
- `$show_song_art` == true` --> result TRUE

Conclusion: 
`$show_song_art` is casted into a boolean variable and cannot be compared to the related text values by a string comparison.
Would you agree or did I do something wrong in my tests? Any missing setting perhaps?